sentry-core/src/main/java/io/sentry/core/transport/HttpTransport.java

```diff
@@ -296,47 +279,95 @@ public boolean isRetryAfter(final @NotNull String itemType) {
 
-    try (final OutputStream outputStream = connection.getOutputStream();
-        final Writer writer = new BufferedWriter(new OutputStreamWriter(outputStream, UTF_8))) {
+    try (final ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
+        final OutputStream outputStream = connection.getOutputStream();
+        final GZIPOutputStream gzip = new GZIPOutputStream(byteOut);
+        final Writer writer = new BufferedWriter(new OutputStreamWriter(gzip, UTF_8))) {
       serializer.serialize(envelope, writer);
 
-      // need to also close the input stream of the connection
-      connection.getInputStream().close();
+      writeAndFlushStreams(connection, byteOut, outputStream, gzip);
       options.getLogger().log(DEBUG, "Envelope sent successfully.");
       return TransportResult.success();
     } catch (IOException e) {
-      try {
-        responseCode = connection.getResponseCode();
-
-        // TODO: 403 part of the protocol?
-        if (responseCode == HttpURLConnection.HTTP_FORBIDDEN) {
-          options
-              .getLogger()
-              .log(DEBUG, "Envelope was rejected by the Sentry server due to a filter.");
-        }
-        logErrorInPayload(connection);
-        return TransportResult.error(responseCode);
-      } catch (IOException responseCodeException) {
-        // this should not stop us from continuing. We'll just use -1 as response code.
-        options
-            .getLogger()
-            .log(WARNING, "Failed to obtain response code while analyzing event send failure.", e);
-      }
-
-      logErrorInPayload(connection);
-      return TransportResult.error(responseCode);
+      final TransportResult errorResult = getResultFromErrorCode(connection, e);
+      responseCode = errorResult.getResponseCode();
+      return errorResult;
     } catch (Exception e) {
-      options
-          .getLogger()
-          .log(WARNING, "Failed to obtain error message while analyzing event send failure.", e);
-      return TransportResult.error(-1);
+      return getDefaultErrorResult(e);
     } finally {
       updateRetryAfterLimits(connection, responseCode);
 
       connection.disconnect();
     }
   }
 
+  /**
+   * Writes and flushes the streams
+   *
+   * @param connection the HttpURLConnection
+   * @param byteOut the ByteArrayOutputStream
+   * @param outputStream the OutputStream
+   * @param gzip the GZIPOutputStream
+   * @throws IOException throws IOException if error while manipulating the streams
+   */
+  private void writeAndFlushStreams(
+      final @NotNull HttpURLConnection connection,
+      final @NotNull ByteArrayOutputStream byteOut,
+      final @NotNull OutputStream outputStream,
+      final @NotNull GZIPOutputStream gzip)
+      throws IOException {
+    gzip.finish();
+    outputStream.write(byteOut.toByteArray());
+    outputStream.flush();
+
+    // need to also close the input stream of the connection
+    connection.getInputStream().close();
```

Comment:
H: Should we take the `InputStream` here instead of the connection?

if it throws above, do we leak resources here? DO we need this in a `finally` block?

Reply:
this method is called inside a https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html
`OutputStream` implements `Closeable` so it's not an issue, but it does make sense to take the `InputStream` and I'll do it.
I've used the IDE to extract this to a method so it went "as before"